daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -2295,6 +2332,20 @@ abstract class FunctionCallBase(
     res
   }
 
+  protected def checkArgArrayOrOptional(): Unit = {
```

Comment:
If this function is going to be in FunctionCallBase, then it would be useful if it could be generic enough to work for any possible function. Right now, it only looks at expression.head, so it's only useful for functions that require array/optional as the first function parameter. I'd suggest changing it to allow it to pass in an index that can be used to pick which expression it checks. The fn:count function would call it like `checkArgArrayOrOption(0)`.

Alternatively, move this logic into FunctionCallArrayBase, since it's likely the only thing that need will need this kindof check. It already has `arrPath.isPathToOneWholeArray`, maybe that just needs to be updated to require this?

That said, I'm not sure why arrayPath.isPathToOneWholeArray in FunctionCallArrayBase doesn't already result in an SDE. Seems like that's already doing the check we want. I wonder if there's just a bug there and that's not getting checked correctly, and we don't actually need this new checkArgArrayOrOptional check?

Reply:
I took another scan at this, and it looks like `isTypeCorrect` (which already calls isPathToOneWholeArray) is never actually used. I suggest we just delete this val and the couple of references to it, and then move your `checkArgArrayOrOptional` function to `FunctionCallArrayBase`. It can then use arrPath (instead of expression.head) to verify that is goes to a whole array. And then update implementations of FunctionCallArrayBase to call this function, since they should all go to a whole array.